### ISSUE TITLE
Getting Error While using Multi GPU with DataParallel and ModelParallel with nn.ParallelTable

### ISSUE THREAD

**Issue by jbansal** [torch/cutorch#260]
Hi All,
I am creating Model which can run using modelParallel to use power of multiple GPUs.
When I set GPU =1 then my model is working fine, but when I set GPU >1 , then I am getting following error.

/usr/local/share/lua/5.1/cutorch/init.lua:21: /usr/local/share/lua/5.1/nn/Linear.lua:51: Assertion `THCudaTensor_checkGPU(state, 4, r_, t, m1, m2)' failed.  at /home/jitendra/cutorch/lib/THC/THCTensorMathBlas.cu:79
stack traceback:
        /usr/local/share/lua/5.1/fb/util/error.lua:76: in function </usr/local/share/lua/5.1/fb/util/error.lua:72>
        [C]: in function 'error'
        /usr/local/share/lua/5.1/cutorch/init.lua:21: in function 'withDevice'
        /usr/local/share/lua/5.1/fbcunn/AbstractParallel.lua:134: in function 'updateOutput'
        /usr/local/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
        /usr/local/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
        Main_MultiGPU.lua:87: in function 'opfunc'
        /usr/local/share/lua/5.1/optim/sgd.lua:43: in function 'sgd'
        Main_MultiGPU.lua:117: in function <Main_MultiGPU.lua:66>
        [C]: in function 'real_xpcall'
        /usr/local/share/lua/5.1/fb/util/error.lua:85: in function 'xpcall'
        /usr/local/share/lua/5.1/threads/threads.lua:171: in function 'dojob'
        /usr/local/share/lua/5.1/threads/threads.lua:259: in function 'synchronize'
        Main_MultiGPU.lua:190: in main chunk
        [C]: in function 'dofile'
        [string "_RESULT={dofile "Main_MultiGPU.lua" }"]:1: in main chunk
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/trepl/init.lua:630: in function 'repl'
        /usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
        [C]: at 0x00406260

I am using following Below Model configuration :

require 'cudnn'
require 'fbcunn'

local features1 = nn.Sequential()
features1:add(cudnn.SpatialConvolution(3,48,5,5,2,2))       -- 64 -> 30
features1:add(cudnn.ReLU(true))
features1:add(cudnn.SpatialMaxPooling(2,2,2,2))                   -- 30 ->  15
features1:add(cudnn.SpatialConvolution(48,64,3,3,2,2))       --  15 -> 7
features1:add(cudnn.ReLU(true))
features1:add(cudnn.SpatialMaxPooling(2,2,2,2))                   --  7 ->  3
features1:add(cudnn.SpatialConvolution(64,128,1,1,1,1))      --  3 ->  3 
features1:add(nn.Reshape(128*3*3))

local features2 = nn.Sequential()
features2:add(cudnn.SpatialConvolution(3,48,5,5,2,2))       -- 256 -> 126
features2:add(cudnn.ReLU(true))
features2:add(cudnn.SpatialMaxPooling(2,2,2,2))                   -- 126 ->  63
features2:add(cudnn.SpatialConvolution(48,64,3,3,2,2))       --  63 -> 30
features2:add(cudnn.ReLU(true))
features2:add(cudnn.SpatialMaxPooling(2,2,2,2))                   --  30 ->  15
features2:add(cudnn.SpatialConvolution(64,128,1,1,1,1))      --  15 ->  6

features2:add(nn.Reshape(128*3*3))
nGPU=opt.nGPU

if nGPU > 1 then
	assert(nGPU <= cutorch.getDeviceCount(), 'number of GPUs less than nGPU specified')
	local features_First = features1
	local features_Second = features2
	require 'fbcunn'
	features1 = nn.DataParallel(1)
	features2 = nn.DataParallel(1)
	for i=1,nGPU do
		cutorch.withDevice(i, function()
			features1:add(features_First:clone())
		end)
		cutorch.withDevice(i, function()
			features2:add(features_Second:clone())
		end)
	end
	features1.gradInput = nil
	features2.gradInput = nil
end
]]
local LinearModel = nn.Sequential()

  local branch1
   if nGPU == 1 then
      branch1 = nn.Concat(2)
   else
      branch1 = nn.ModelParallel(2)
   end
   for i=1,nGPU do
      local s = nn.Sequential()
      s:add(nn.Linear(128*3*3*2, 512/nGPU))
      s:add(nn.Tanh())
      branch1:add(s)
   end
   LinearModel:add(branch1)
   local branch2
   if nGPU == 1 then
      branch2 = nn.Concat(2)
   else
      branch2 = nn.ModelParallel(2)
   end
   for i=1,nGPU do
      local s = nn.Sequential()
      s:add(nn.Linear(512, 16/nGPU))
      s:add(nn.Tanh())
      branch2:add(s)
   end
   LinearModel:add(branch2)
   LinearModel:add(nn.Linear(16, 1))  
   

model = nn.Sequential()
local Parallelmodel =nn.ParallelTable():add(features1):add(features2)
model=model:add(Parallelmodel):add(nn.JoinTable(2)):add(LinearModel)

print("Model Loaded")

I am passing the data to this model as follows:

Predictions =nn.Sigmoid():cuda():forward(model:forward({com_inputA,com_inputB})) --com_inputA,com_inputB is the image pixel values (nChannels*width*height)

In this model I am passing two images in Parallel to model which will get features of images and then pass this features to linear model to get if both images are similar or not.

is it due to nn.ParallelTable()

Please help me to find out the issue.

Regards,
Jitendra Bansal

**Comment by fmassa** [comment]
@jbansal could we continue the discussion [here](https://groups.google.com/forum/#!topic/torch7/CBSTDLoIEio) ? I'll help you there (even though I've never used `fbcunn`).

**Comment by soumith** [comment]
@jbansal next time, please make your example smaller. It is hard to read through so much code. Also, hopefully you got the getParametersByDevice branch of nn as specified in fbcunn's instructions.

**Comment by jbansal** [comment]
Hi Soumith,
I am getting following error even for simple model also.
I am using multi threading as explained in fbcunn imagenet example. 
I am using function call for loading models in different thread and then do model:forward() and backward operations.

When I set GPU=1 then it is working fine. But when I do GPU=2 or more , I am getting following error.


/usr/local/share/lua/5.1/threads/threads.lua:264:
[thread 1 callback] /usr/local/share/lua/5.1/nn/Linear.lua:51: Assertion `THCudaTensor_checkGPU(state, 4, r_, t, m1, m2)' failed.  at /tmp/luarocks_cutorch-scm-1-8349/cutorch/lib/THC/THCTensorMathBlas.cu:79
stack traceback:
        /usr/local/share/lua/5.1/fb/util/error.lua:76: in function </usr/local/s
hare/lua/5.1/fb/util/error.lua:72>
        [C]: in function 'addmm'
        /usr/local/share/lua/5.1/nn/Linear.lua:51: in function 'updateOutput'
        /usr/local/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
        ...endra/Desktop/WhatTheFont_2/SynathData/LoadModelData.lua:70: in function 'opfunc'
        /usr/local/share/lua/5.1/optim/sgd.lua:43: in function 'sgd'
        ...endra/Desktop/WhatTheFont_2/SynathData/LoadModelData.lua:116: in function 'LoadMultipleGPU'

to understand better, Please find the below code (I tried to make it simple and short):

--Model--
model =nn.Sequential()
model:add(nn.Linear(10,2))
model:add(nn.Tanh())
model:add(nn.Linear(2,1))

--Multi threading call

donkeys:addjob(
			function()
				local GPUID=imgCount%opt.nGPU+1
				local loss =LoadMultipleGPU(GPUID)
				return loss
			end,
			callbackFunc
		if imgCount % 5 == 0 then
			donkeys:synchronize()
			collectgarbage()
		end


-- Model function call


function LoadMultipleGPU(GPUID)
	collectgarbage()
	cutorch.setDevice(GPUID)
	
	model:cuda()
	criterion:cuda()
	parameters, dl_dx = model:getParameters()
	local Loss=0
	local comInput = torch.rand(1,10)
	local comOut=torch.FloatTensor(1,1):fill(1)
	
	local feval = function(x)
		if x ~= parameters then
			x:copy(parameters)
		end
		-- Zero the gradients 
		dl_dx:zero()
		local Predictions =nn.Sigmoid():cuda():forward(model:forward(comInput:cuda()))
		local deltaClone =torch.FloatTensor(1,1):fill(0.5)
		local LossVector =criterion:forward(deltaClone:cuda(),comOut:cuda())
		Loss=Loss+LossVector:sum()
		local gradInput = criterion:backward(deltaClone:cuda(),comOut:cuda())
		model:backward(comInput:cuda(),gradInput)
		dl_dx  = dl_dx:div(2)
		return Loss, dl_dx
	end
	_,fs = optim.sgd(feval,parameters,sgd_params)
	return Loss
end
